PSReadLine/UndoRedo.vi.cs

```diff
@@ -3,24 +3,36 @@
 --********************************************************************/
 
 using System;
+using System.Collections.Generic;
 
 namespace Microsoft.PowerShell
 {
     public partial class PSConsoleReadLine
     {
-        private class GroupUndoHelper
+        /// <summary>
+        /// Encapsulates state and behaviour for an edit group.
+        ///
+        /// An edit group is a sequence of distinct internal changes
+        /// to the buffer that are considered to be a single user-facing
+        /// change for the purpose of the undo/redo actions.
+        /// </summary>
+        internal struct GroupUndoHelper
         {
             public Action<ConsoleKeyInfo?, object> _instigator;
             public object _instigatorArg;
 
-            public GroupUndoHelper()
-            {
-                _instigator = null;
-                _instigatorArg = null;
-            }
-
             public void StartGroup(Action<ConsoleKeyInfo?, object> instigator, object instigatorArg)
             {
+                if (_singleton._editGroupStart != -1)
+                {
+                    // A nested "start" of a group is being made, so we
+                    // need to record the state of the preceding start.
+                    _singleton._groupUndoStates.Push(
+                        new GroupUndoState(_singleton._groupUndoHelper, _singleton._editGroupStart));
+
+                    _singleton._editGroupStart = -1;
+                }
+
                 _instigator = instigator;
                 _instigatorArg = instigatorArg;
```

Comment:
It doesn't seem to be right to override the existing `_instigator` and `_instigatorArg`. Consider `ViReplaceToChar` as an example, it calls `ViInsertWithAppend` internally, which also starts a group with your change, however, when redoing, it should be calling the instigator set by `ViReplaceToChar` instead of the one set by `ViInsertWithAppend`, because the use didn't explicitly trigger `ViInsertWithAppend`.